### DISCUSSION TITLE
AuxVariables in Functions

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#20984] [Q&A General] [answered]
Hello,

I was wondering if there is a way to use an AuxVariable in a function? I'd like to write a function that is dependet on the contact pressure of my system which is stored as an AuxVariable. But the vars/vals functionality in the function-block seems to be restricted to to postprocessors, scalars and other functions. I need the spatial and temporal changes of the contact pressure though, so I tried a NodalValueSampler Vector Postprocessor, but the function-block seems to not check for those. 

Is there another way to do this?

**Comment by Flolaffel** [comment, marked as answer]
Ok, I just found the VectorPostprocessorFunction which seems to do exactly what I was looking for

**Comment by GiudGiud** [comment]
Typically not. The only function I know of that can use variables instead of just location information is the SolutionFunction
https://mooseframework.inl.gov/moose/source/functions/SolutionFunction.html
but that is information from a SolutionUserObject

**Reply by Flolaffel** [reply]
Well my procedure is: 
1. Use a NodalValueSampler to read the contact pressure on a boundary
2. I now have a postprocessor that contains the values of my AuxVariable
3. Use a VectorPostprocessorFunction that reads the contact pressure from the postprocessor

But there seems to be a problem here. While the NodalValueSampler updates its values every timestep, the VectorPostprocessorFunction does not. So I have multiple steps where the function doesn't seem to recognize the change in contact pressure. Do you have an idea why this could be happening?

**Reply by GiudGiud** [reply]
This is likely because of the order of execution of each object. There is no cross-system dependency resolution.
You are using an AuxKernel afterwards too right?

You can remedy this to some extent by setting execute_on flags. The linear/nonlinear ones are about as often as you can run each object.  see some docs here
https://mooseframework.inl.gov/source/interfaces/SetupInterface.html

The VPPfunction reads directly from the VPP and performs an interpolation so if the Sampler is updated the VPPFunction is necessarily up to date.

**Reply by Flolaffel** [reply]
No, there is no AuxKernel involved.

I tried the execute_on flags. VPPfuncion is executed on linear by default. Changing the execution of the VPP to linear also didn't solve the problem.

Plus the VPP doesn't seem to be causing the problem. It's the VPPfunction that seems to fail to read the VPP.

**Reply by GiudGiud** [reply]
how do you diagnose that the function isnt being updated?
In what object are you using it?

You can try commenting out:
```
  if (now != _last_update)
  {
```
in valueInternal in VPPFunction.C, it's the only thing that prevents updating under very narrow conditions which I dont think you would be meeting.

**Reply by Flolaffel** [reply]
I'm using the VPPfunction for a DirichletBC and I'm printing it into csv with a LineFunctionSampler. 

So firstly I see in the exodus file when my BC doesn't change between timesteps and secondly I see the csv output which corresponds to my BC and isn't updated either for several timesteps.

**Reply by Flolaffel** [reply]
Commenting out the if-statement actually worked. Thank you!

**Reply by GiudGiud** [reply]
Ah ok there must be a bug with it. I ll fix it. Thanks for trying it out

**Reply by GiudGiud** [reply]
ok I have the fix in a PR. Should be in in a couple weeks at most
